### PR TITLE
Rewrite domainDescriptor test.

### PR DESCRIPTION
**Pull request by beauby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #303 Optimize pinned memory allocation in CudaBasic.
* #302 Fix synchronization in CUDA tests.
* #289 Add control connection for CudaBasic.
* #288 Refactor channel client-server tests.
* **#287 Rewrite domainDescriptor test.**

Differential Revision: [D26514270](https://our.internmc.facebook.com/intern/diff/D26514270)

### CODE REVIEW

File: tensorpipe/test/channel/channel_test.cc [hunk 1]
```diff
@@ -17,23 +17,21 @@ using namespace tensorpipe::channel;
...
+    auto peerGroup = helper->makePeerGroup();
+    peerGroup->spawn(
+        [&] {
+          std::shared_ptr<Context<TBuffer>> context1 =
+              helper->makeContext("ctx1");
+          std::shared_ptr<Context<TBuffer>> context2 =
+              helper->makeContext("ctx2");
+          EXPECT_FALSE(context1->domainDescriptor().empty());
+          EXPECT_FALSE(context2->domainDescriptor().empty());
+          EXPECT_EQ(context1->domainDescriptor(), context2->domainDescriptor());
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.cc:32]: Just noticed (and unrelated to this diff): do we want to now change this to check that `context1->canCommunicateWithRemote(context2->domainDescriptor())`?

File: tensorpipe/test/channel/channel_test.cc [hunk 2]
```diff
@@ -17,23 +17,21 @@ using namespace tensorpipe::channel;
...
+        },
+        [] {});
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.cc:34]: I see you removed the `done` and `join` calls. I don't remember what they were for. Did they only ensure there was a "barrier"? (in which case I agree we don't need them)

**Review by lw — APPROVED** [review]

File: tensorpipe/test/channel/channel_test.cc (hunk 2 above)
Comment (beauby) [reply tensorpipe/test/channel/channel_test.cc:34]: Yeah so join and done are just a messaging-based barrier on the peer group. The peer group itself joins both threads/processes, but the barrier is/was needed to keep objects alive on the side that would join first.

File: tensorpipe/test/channel/channel_test.cc (hunk 1 above)
Comment (beauby) [reply tensorpipe/test/channel/channel_test.cc:32]: Good question. In its current form, this test somewhat ensures that the domain descriptor is the same in subsequent calls. I'm not sure which property we want to test.

Comment (lw) [reply tensorpipe/test/channel/channel_test.cc:32]: Well we can't really check for any "deep" semantic meaning of the descriptor (like, that it compares true or false in the various scenarios), as that varies for each channel and it requires specific test setups. So I guess we're mainly checking for trivial properties, and "reflexivity" seems a good one. However, I also now think that ensuring that all contexts started in the same conditions should produce the same descriptor. So maybe we should both check for equality, and that `canCommunicateWithRemote` is true?

**Pull request closed without merging**